Feature/canonical pdfs

miggol opened the pull request:
This addresses #446, releasing us from the clutches of old proposal compatibility. (I wish)

Creating this PR as a draft because I want to test some business logic that I moved around.

miggol commented:
This is now ready for review!

miggol commented:
Absolutely, I would appreciate your review as well @tymees. Take your time

tymees commented on reviews/views.py:
```diff
@@ -354,14 +354,13 @@ def form_valid(self, form):
-            ]
-            proposal.date_reviewed = timezone.now()
-            proposal.save()
+            proposal.mark_reviewed(form.instance.continuation)
```
Github won't let me comment on the code below, this comment is about lines 364-372

The rest of this method looks so nice and clean now, but that part still follows the old doctrine. From what I can tell, it would be safe to consolidate that entire section in ``start_review_route`` (and _maybe_ also move that into a model) 

The same argument can be said for other views in this file, but that's getting a bit too far from the scope of this PR. It might be nice to create a seperate issue for this as well, it's a nice starter-issue for our new colleague when they start.

miggol replied:
I totally agree, but it's just out of scope for this PR. I don't feel like writing tests for all that right now.

tymees reviewed (changes requested):
I like what I see, and as usual have added some comments. 

However, there is still 2 issues I cannot pin to specific code in the PR:

``generate_pdf`` is being called all over the place using the old signature:
![image](https://user-images.githubusercontent.com/7061963/223874214-cc87ed2b-0645-4cbb-b1d6-8e6544442f23.png)
If we are mucking about with PDF generation, we really should clean those up. 
(I should've picked this up in the PR where we switched the PDF generation method, but well, here we are...)

The second issue is actually something that's been overlooked for years; the PDF does not update when the secretary changes the documents manually. (``StudyUpdateAttachments`` in ``study_views.py``).

This is a bit of a breaking problem if we're going to start using the prerendered PDF as the 'only factual representation of a proposal after review'. 

Obviously, we could just add regenerate the PDF when that happens. And short-term, I think that is the safest best. It's still a bit of a problem as it might happen after we change the PDF template, but as the chance of that happening being slim I think it's good enough for now.

Long term, I think we should move to using links that don't refer to files on the filesystem and use deterministic urls. (e.g. /documents/1/1/blaat/ for the 'blaat' file for proposal 1, study 1). That way the url in the PDF does not have to change when the file changes.
I suspect this will happen when we switch over to the new documents system. (As ``cdh.files`` requires custom Views to serve files anyway).


... And here I was telling myself I'd keep my comments small lol, my apologies for the multiple walls of text. Also, my brain is still a little fried, so please be gentle if I said something stupid ;)

miggol commented on reviews/tests.py:
```diff
@@ -267,3 +300,168 @@ def test_auto_review_task(self):
         self.assertEqual(len(reasons), 1)
+
+
+class BaseViewTestCase(BaseReviewTestCase):
```
The reason is that `TestCase` does not follow the Mixin design pattern. You could easily end up calling a non-existent `super().setUp()` because `setUp()` isn't defined inthe base class.

That's probably solvable with a stub base class, but I´d rather do the generalizing work when we need it so we actually know what problems need solving. I've renamed the class to `BaseReviewViewTestCase` to better reflect that it's more of a bespoke thing for reviews only.

tymees replied:
> I've renamed the class to BaseReviewViewTestCase to better reflect that it's more of a bespoke thing for reviews only. (For now, I totally support using this more in the future!)

Good enough for now ;) 

However, some of your argumentation for why a mixin wouldn't work doesn't hold up in my eyes:

> The reason is that ``TestCase`` does not follow the Mixin design pattern.

So? The fact that Django's ``TestCase`` doesn't use Mixins doesn't mean it's not a valid design pattern to use here. It only means Django didn't need it. [Django's own tests](https://github.com/django/django/blob/78da5ca0c1f2ab3201f8f6cd629e80d805ea023d/tests/test_utils/test_serializemixin.py) do use them ;)

> You could easily end up calling a non-existent ``super().setUp()`` because ``setUp()`` isn't defined inthe base class.

I don't see how. ``setUp()`` is defined on [topmost base class](https://docs.python.org/3/library/unittest.html#unittest.TestCase) of Django's ``TestCase``. 

And you're never going to use a Mixin intended for ``TestCase`` outside of a ``TestCase``; just like you would never use a ``View`` mixin outside of a ``View``. (Well, you obviously can, but shouldn't).

miggol replied:
I can't recall the problem I ran into earlier, but the BaseViewTestCase and all other tests now follow the mixin pattern. The other tests had to be adjusted accordingly.

tymees replied:
A more general location would maybe be DSC? The mixin is general enough to use in other apps as well.

Otherwise, the `main` app is (in my eyes) a good place for app-wide code. Lots of view mixins for example already live there

miggol commented on reviews/tests.py:
```diff
@@ -267,3 +300,168 @@ def test_auto_review_task(self):
+
+    def get_view_path(self):
+        pk = self.review.pk
+        return f"/reviews/close/{pk}/"
```
Because I want the test to fail if the literal URL changes.

Most likely scenario is that such a URL change would be intentional. But it's a good canary for unintentional URL changes. Those would slip under the radar if we used `reverse()` here. Probably not so important for this view specifically, but good practice for tests in general IMO.

Added a comment to reflect this.

tymees replied:
I can see where you're coming from, but I still think it's strange;

> Because I want the test to fail if the literal URL changes.

If you want to test something, then write a test for that. ;)
The thing is, this method is not a test itself. Thus, seemingly unrelated tests will fail if the url changes, which would make me mildly annoyed even if it's documented in the testcase itself.

If your objective is to make sure the URL doesn't change with a test, I would prefer to make it an actual dedicated testcase. (And use reverse everywhere else). I think it's better practice to keep your tests as focused as possible and avoid other problems from affecting a specific testcase. That way it's more obvious why we have a failing test, thus accomplishing the goal of testing more effectively. 

Also, if we were to have a specific ``TestCase`` class for URLs, it would give us an oppertunity to add testcases for URLs that are used outside the app (or in emails). Because you're right, we should be checking that ;)

miggol replied:
On second thought, it actually does make sense to test the `reverse()` functionality for this view as well. That would be a canary for an unintentional view name (i.e. `reviews:close`) change. But I'm a little conflicted on whether that belongs in this test or in the test for the views that actually call `reverse("reviews:close")`. Will think about it some more

miggol replied:
Added a separate routing test to cover both cases

miggol commented on proposals/models.py:
```diff
@@ -560,6 +562,39 @@ def latest_review(self):
+
+    def mark_reviewed(self, continuation, time=None):
+        """Finalize a proposal after a decision has been made."""
+        if not time:
```
Good catch! Old habits...

miggol commented on proposals/models.py:
```diff
@@ -560,6 +562,39 @@ def latest_review(self):
+        on request. Proposals that have already been decided on should
+        rely on a PDF generated at time of review, so that the PDF
+        generation templates can evolve without breaking older proposals."""
+        return self.status_review is not None
```
> I do think this is a nicer place to handle that edgecase, as it's safer to assume our Model will persist than our View.

That makes sense, perhaps we could add a safe `get_recommended_pdf()` function to the model. In my eyes, the singular purpose of this specific view is to answer the question "Is this proposal in a state for which a canonical pdf should be used." I don't want to add a conditional to that.

miggol commented on proposals/mixins.py:
```diff
@@ -106,7 +106,7 @@ def get_pdf_response(self, context, dest=None, **response_kwargs):
 
         # find the template and render it.
-        template = get_template(self.template_name)
+        template = get_template(self.get_template_names()[0])
```
Empty lists and non-lists are now handled more gracefully

miggol commented:
Alright, I've made changes and brought this branch back up to speed with `develop`. This is ready for re-review.

* Tests now follow the mixin pattern.
* `generate_pdf()` and `pdf_template_name()` now also reside in the Proposal model. 
* Addressed comments

Achnowledgements:
* Injecting the object into the ProposalAsPDF view manually still isn't ideal, but good enough for now. In the future I might want to move to using a RequestFactory like in the tests.
* Plenty of proposal and review logic still resides in the views. But this is a start and we now have better ways to test while moving forward.

Discussion question:
* Should we make it so a new PDF *can not* be saved if a proposal is in a state where it defers to its canonical PDF? This would prevent old proposals from having their holy reference PDFs overwritten.

tymees commented on proposals/mixins.py:
```diff
@@ -106,7 +107,17 @@ def get_pdf_response(self, context, dest=None, **response_kwargs):
-        template = get_template(self.template_name)
+        template_names = self.get_template_names()
+
+        if not (type(template_names) is tuple or list):
```
I don't think this works as intended?
The 'or' here is not part of the 'is' statement, but rather a 'what if the previous thing is false'. 
I did a small sanity check, and it seems to agree with me:
```
>>> x = []
>>> type(x) is tuple or list
<class 'list'>
>>> x = ()
>>> type(x) is tuple or list
True
```

I'd use this instead: 
```python
if type(template_names) not in [tuple, list]:
```
Alternatively:
```python
from collections.abc import Iterable
[..]
if not isinstance(template_names, Iterable):
```

miggol replied:
Good catch. The logic made no sense. It now checks for an instance of `Iterable`.

tymees commented on proposals/views/proposal_views.py:
```diff
@@ -541,36 +542,57 @@ def get_context_data(self, **kwargs):
+
+    def get_template_names(self):
+        """Determine the correct PDf template for given proposal"""
+        self.template_name = 'proposals/proposal_pdf.html'
```
This first line seems redundant?

miggol replied:
This vestigial appendage has been dealt with

tymees commented on proposals/views/proposal_views.py:
```diff
@@ -541,36 +542,57 @@ def get_context_data(self, **kwargs):
+        # Else, continue with generation
+        return super().get(request, *args, **kwargs)
+
+    def get_object(self):
+        """If we already have an object set, use that.
+        This can happen if this view is being called from generate_pdf()
+        rather than by Django."""
+        if not hasattr(self, "object"):
+            self.object = super().get_object()
```
I get that my huge comment about this method is a bit annoying, and mostly out-of-scope for the PR, but there was an actual small thing that is still relevant in there:

The signature is incompatible with the original method (which has an additional argument). It would be nicer to a) copy over the original signature or b) add *args, **kwargs to this method and pass those in the super().get_object() call.

miggol replied:
Ah yes, that part was indeed in need of fix. I have gone the `*args, **kwargs` route now.

tymees reviewed (changes requested):
Looks very good!

I still have 3 very minor points (of which only the first one is serious enough for me not to click the approve button). 

As for your comment:
> Injecting the object into the ProposalAsPDF view manually still isn't ideal, but good enough for now. In the future I might want to move to using a RequestFactory like in the tests.

Fine with me, this PR is already big enough as-is :)

> Should we make it so a new PDF can not be saved if a proposal is in a state where it defers to its canonical PDF? This would prevent old proposals from having their holy reference PDFs overwritten.

Yes, I think that would be a good idea.

Maybe add a parameter 'force-overwrite' that defaults to False; there are some cases we might actually want to force it anyways. (The secretary updating documents during approval comes to mind, as well as the regerate_pdf command)

tymees reviewed (approved):
It would be nice if the new 'force_overwrite' was used in the management command, possibly as an additional argparse flag

However, it's not mandatory, as we can do the same thing using the Django shell if need be. So, LGTM

miggol commented:
I've improved the management command with your suggestion, @tymees. Merging now!
